File: scala/sublist/src/main/scala/Sublist.scala
class Sublist {
  def sublist[A](x: List[A], y: List[A]): Sublist.SublistType =
    if (x.equals(y)) Sublist.Equal
    else if (x.containsSlice(y)) Sublist.Superlist
    else if (y.containsSlice(x)) Sublist.Sublist
    else Sublist.Unequal
}

object Sublist {
  sealed abstract class SublistType
  case object Equal extends SublistType
  case object Unequal extends SublistType
  case object Sublist extends SublistType
  case object Superlist extends SublistType
}

File: fsharp/sublist/Sublist.fs
module Sublist

type SublistType = Equal | Sublist | Superlist | Unequal

let private isSublist (list1: List<'T>) (list2: List<'T>) =
    list1.IsEmpty || List.windowed list1.Length list2 |> List.exists (fun window -> window = list1)

let sublist (list1: List<'T>) (list2: List<'T>) =
    if list1.Length < list2.Length then
        if isSublist list1 list2 then Sublist else Unequal
    elif list1.Length > list2.Length then
        if isSublist list2 list1 then Superlist else Unequal
    else
        if list1 = list2 then Equal else Unequal
